### PR TITLE
Aarch64 peephole using csinc

### PR DESCRIPTION
**Pull request by swalk-cavium**
This change introduces an Aarch64 specific pseudo instruction which
is a variation of the conditional move.  One pattern seen in the OOPSLA'14
benchmark tests involved the expression  like this in C

    (cond) ? exp : 1;

The hard-coded zero register can be used with the csinc instruction to
create the 1.

Before
======
     0x146010d0  39c0e361             ldrsb w1, [x27, #56]
     0x146010d4  72001c3f             tst w1, #0xff
     0x146010d8  52800022             movz w2, #0x1   //<<---
     0x146010dc  1a810041             csel w1, w2, w1, eq

After
=====
      0x362011d0  38e16b61              ldrsb w1, [x27, x1] 
      0x362011d4  72001c3f              tst w1, #0xff
      0x362011d8  1a9f1421              csinc w1, w1, wzr, ne  

This saves 1 instruction per instance.

The standard regression suite was run with 6 option sets.  No new regressions
were introduced.

### CODE REVIEW

**Comment by swalk-cavium** [comment]
@mxw - Hi Max, Can you review this one?

File: hphp/runtime/vm/jit/vasm-instr.h [hunk 1]
```diff
@@ -1101,6 +1105,10 @@ struct shlq { Vreg64 s, d; VregSF sf; Vflags fl; }; // uses rcx
 /*
  * arm intrinsics.
  */
+struct csoneb { ConditionCode cc; VregSF sf; Vreg8 t, d; };
```
Comment (cmuellner) [inline hphp/runtime/vm/jit/vasm-instr.h]: Why not a real csinc{} and use `PhysReg(vixl::zr)` as the second source register?
`Conditional Select Increment returns, in the destination register, the value of the first source register if the condition is TRUE, and otherwise returns the value of the second source register incremented by 1.`

File: hphp/vixl/a64/macro-assembler-a64.h [hunk 2]
```diff
@@ -422,7 +421,6 @@ class MacroAssembler : public Assembler {
     assert(allow_macro_instructions_);
     assert(!rd.IsZero());
     assert(!rn.IsZero());
```
Comment (cmuellner) [inline hphp/vixl/a64/macro-assembler-a64.h]: You might want to remove the assertion `assert(!rn.IsZero());` as well.
Upstream vixl has this code:

```
void MacroAssembler::Csinc(const Register& rd,
                           const Register& rn,
                           const Register& rm,
                           Condition cond) {
  DCHECK(allow_macro_instructions_);
  DCHECK(!rd.IsZero());
  DCHECK((cond != al) && (cond != nv));
  csinc(rd, rn, rm, cond);
}
```

File: hphp/runtime/vm/jit/vasm-instr.h (hunk 1 above)
Comment (swalk-cavium) [inline hphp/runtime/vm/jit/vasm-instr.h]: I thought we couldn't have PhysRegs in the vasm-instrs.  

The peephole flips the condition code if necessary and only creates the
pseudo if semantically correct.

Comment (cmuellner) [inline hphp/runtime/vm/jit/vasm-instr.h]: We are doing the PhysReg() trick in the lowering of storeNi{} in vasm-arm.cpp.

Comment (cmuellner) [inline hphp/runtime/vm/jit/vasm-instr.h]: Also vasm-fold-imms.cpp is full of that. E.g.:
```
  void fold(storeb& in, Vinstr& out) {
    if (zero_imm(in.s)) out = storeb{PhysReg(vixl::wzr), in.m};
  }
```

Comment (swalk-cavium) [inline hphp/runtime/vm/jit/vasm-instr.h]: In vasm-fold-imms.cpp they're in a 'namespace arm'.  The other PhysRegs are either in
target specific or register specific files.   The only target specific code in vasm-simplify are the tests against arch().  There are Vinstrs labeled as intrinsics, so I thought it safe to
create one of those.  You have to pick something which avoids using the reg in the initial
move otherwise it won't be eliminated as dead code.  The pseudo instruction seemed like the right place to emit the xzr.

**Comment by mxw** [comment]
Whoops, sorry for the tag spam; I had the wrong PR.

File: hphp/runtime/vm/jit/vasm-instr.h (hunk 1 above)
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-instr.h]: I just split out the ARM-specific simplifications into a separate file (in 16710e2), so you can move all your new simplifications there and just define `csinc{}`.

File: hphp/runtime/vm/jit/vasm-simplify.cpp [hunk 3]
```diff
@@ -196,6 +196,37 @@ bool simplify(Env& env, const setcc& vsetcc, Vlabel b, size_t i) {
   });
 }
 
+template <typename Reg>
+bool operand_one(Env& env, Reg op) {
```
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: Why did you template this function?  Why can't you just use `Vreg`?

File: hphp/runtime/vm/jit/vasm-simplify.cpp [hunk 4]
```diff
@@ -196,6 +196,37 @@ bool simplify(Env& env, const setcc& vsetcc, Vlabel b, size_t i) {
...
+// Reduce use of immediate one possibly removing def as dead code.
+// Specific to ARM using hard-coded zero register.
+template <typename Out, typename Inst>
+bool cmov_fold_one(Env& env, const Inst& inst, Vlabel b, size_t i) {
+  if (arch() != Arch::ARM) return false;
+  if (operand_one(env, inst.f)) {
+    return simplify_impl(env, b, i, [&] (Vout& v) {
+      v << Out{inst.cc, inst.sf, inst.t, inst.d};
+      return 1;
+      });
```
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: nit: indentation

File: hphp/runtime/vm/jit/vasm-simplify.cpp [hunk 5]
```diff
@@ -196,6 +196,37 @@ bool simplify(Env& env, const setcc& vsetcc, Vlabel b, size_t i) {
...
+    return simplify_impl(env, b, i, [&] (Vout& v) {
+      v << Out{inst.cc, inst.sf, inst.t, inst.d};
+      return 1;
+      }); 
+  }
+  if (operand_one(env, inst.t)) {
+    return simplify_impl(env, b, i, [&] (Vout& v) {
+      v << Out{ccNegate(inst.cc), inst.sf, inst.f, inst.d};
+      return 1;
+      });
```
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: nit: indentation

File: hphp/runtime/vm/jit/vasm-simplify.cpp (hunk 3 above)
Comment (swalk-cavium) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: @mxw - I'm not sure I understand your comment about csinc{} above.  The csoneb still has
to go through the rest of the passes to get real registers.  I don't think you can make it opaque to the other passes.  I can move the actual peep into the new arm simplify file.

Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: @swalk-cavium—What I mean is that you can do as @cmuellner suggested and implement `csinc*{}` instead of `csone*{}`—you can just specify the `vixl::wzr` operand explicitly.

Comment (swalk-cavium) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: @mxw - Ok, got it.  I thought you were suggesting something else.  Having a separate file will make the boundaries cleaner.

File: hphp/runtime/vm/jit/vasm-simplify-arm.cpp [hunk 6]
```diff
@@ -33,6 +35,57 @@ bool simplify(Env&, const Inst& inst, Vlabel b, size_t i) { return false; }
...
+      v << Out{ccNegate(inst.cc), inst.sf, PhysReg(vixl::wzr), inst.f, inst.d};
+      return 1;
+    });
+  }
+  return false;
+}
+
+bool simplify(Env& env, const cmovb& inst, Vlabel b, size_t i) {
+  if (cmov_fold_one<csincb>(env, inst, b, i)) return true;
+  return false;
```
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify-arm.cpp]: Change this to just `return cmov_fold_one<csincb>(env, inst, b, i);`, and likewise below.

File: hphp/runtime/vm/jit/vasm-simplify.cpp [hunk 7]
```diff
@@ -120,10 +120,8 @@ bool simplify(Env& env, const setcc& vsetcc, Vlabel b, size_t i) {
   });
 }
 
-/*
- * Fold a cmov of a certain width into a copy if both values are the same
- * register or have the same known constant value.
- */
+// Fold a cmov of a certain width into a copy if both values are the same
+// register or have the same known constant value.
```
Comment (mxw) [inline hphp/runtime/vm/jit/vasm-simplify.cpp]: Can you undo this change?

File: hphp/runtime/vm/jit/vasm-simplify-arm.cpp (hunk 6 above)
Comment (swalk-cavium) [inline hphp/runtime/vm/jit/vasm-simplify-arm.cpp]: @mxw - Sure.  I was expecting other tests, but we can address them at the time.

**Comment by swalk-cavium** [comment]
@max - Hi Max, Updated again.
